Lib/test/test_threading.py

```diff
@@ -976,6 +976,120 @@ def import_threading():
         self.assertEqual(out, b'')
         self.assertEqual(err, b'')
 
+    @cpython_only
+    def test_finalize_daemon_thread_hang(self):
+        # bpo-42969: tests that daemon threads hang during finalization
+        script = textwrap.dedent('''
+            import os
+            import sys
+            import threading
+            import time
+            import _testcapi
+
+            lock = threading.Lock()
+            lock.acquire()
+            def thread_func():
+                try:
+                    _testcapi.finalize_thread_hang(lock.acquire)
+                finally:
+                    # Control must not reach here.
+                    os._exit(2)
+
+            t = threading.Thread(target=thread_func)
+            t.daemon = True
+            t.start()
+            # Sleep to ensure daemon thread is blocked on `lock.acquire`
+            time.sleep(0.1)
```

Comment:
Please don't use time as synchronization: https://pythondev.readthedocs.io/unstable_tests.html#don-t-use-sleep-as-synchronization

Maybe add an event to wait until the thread started.

Reply:
I added a comment about the use of sleep.  The test is designed to succeed even if the time ends up being too short, it just would not be testing the thing we are trying to test in that (rare) case.  I think it is difficult to avoid time-based synchronization for this type of test --- even if we use an event to wait until the thread starts, there is still a possible race between the time it signals the event and the time it becomes blocked on `lock.acquire`.

Reply:
Add an event which is set just before ``_testcapi.finalize_thread_hang(lock.acquire)``. While there will still be a race condition, IMO it will be less likely than using a sleep with a fixed duration.

Reply:
The event exists in the current version of code, this sleep still makes sense to me. It'll give the other thread much more of a chance of being blocked on the lock.acquire before this main thread code proceeds.

this test is overall somewhat of a hack with things like sys.stderr.flush replacing, but it is practical enough for its purpose and can be adjusted if we ever find it causing problems or our internals to have changed enough to warrant testing a different way.